chain.go

```diff
@@ -413,19 +408,12 @@ func (c *chain) fail(failure AssertionFailure) {
 		if c.severity == SeverityError {
 			failure.IsFatal = true
 		}
-
-		context = c.context
-		handler = c.handler
-		reportFailure = true
+		c.failure = &failure
 	}()
 
-	if reportFailure {
-		handler.Failure(&context, &failure)
-
-		if chainValidation {
-			if err := validateAssertion(&failure); err != nil {
-				panic(err)
-			}
+	if chainValidation {
+		if err := validateAssertion(&failure); err != nil {
+			panic(err)
 		}
 	}
```

Comment:
Minor issue: let's move this piece to leave(), after Failure() call.

It's more convenient to panic on ill-formed failure after reporting it, not before, because this way we'll first see the report itself (assertion handler will print failure) and then will see the panic. Otherwise we'll see only panic.